Reload changed auth files when modified on disk

fickludd opened the pull request:
Every 5 seconds the auth and role files are checked for modifications, and reloaded if changed. This is to support clustered native auth, where the files are synchronized using rsync or similar.

changelog: Automatically reload auth files from disk every 5 seconds to enable external changes and externally driven cluster synchronisation

fickludd committed:
Enriched FileSystemAbstraction

* renamed renameFile to move and added CopyOptions
* added lastModifiedTime

OliviaYtterbrink committed:
Change to use FileSystemAbstraction

OliviaYtterbrink committed:
Reload auth and roles files on change

fickludd committed:
Added more and nastier tests for auth file reloading

fickludd committed:
Switched scheduling of auth file reload to JobScheduler

chrisvest commented on community/io/src/test/java/org/neo4j/graphdb/mockfs/DelegatingFileSystemAbstraction.java:
```diff
@@ -19,6 +19,7 @@
  */
 package org.neo4j.graphdb.mockfs;
 
+import java.awt.image.ConvolveOp;
```
What is this?

chrisvest commented on community/io/src/test/java/org/neo4j/graphdb/mockfs/EphemeralFileSystemAbstraction.java:
```diff
@@ -592,6 +613,17 @@ public void truncate( File file, long size ) throws IOException
+        EphemeralFileData data = files.get( canonicalFile( file ) );
+        if ( data == null )
+        {
+            throw new FileNotFoundException( "File " + file + " not found" );
```
This should be `NoSuchFileException`.

fickludd replied:
Could be, but all the other methods in `EphemeralFileSystemAbstraction` return FileNotFoundException when the file data is null. This is only following the pattern.

chrisvest commented:
@fickludd Needs rebase.

chrisvest commented on community/io/src/main/java/org/neo4j/io/fs/DefaultFileSystemAbstraction.java:
```diff
@@ -133,8 +135,13 @@ public void deleteRecursively( File directory ) throws IOException
+            Files.move( from.toPath(), to.toPath(), copyOptions );
+            return true; // will throw if failure
+        }
         return FileUtils.renameFile( from, to );
```
This is pretty ugly. The behaviour, potential exceptions and side-effects change depending on whether there are copy options or not, even if they otherwise should make no functional difference.

Have you tried using `Files.move` always, and checked if it works or breaks on all relevant platforms?

fickludd replied:
It's ugly indeed, but I selected this implementation after looking at the even uglier windows specific bug-fix in `FileUtils.renameFile`. I think the full move over to `Files.move` should happen when more time is available to handle the potential bugs.

fickludd commented:
@chrisvest on it, but found threading bug as well, so takes a little time to straighten out

fickludd committed:
Removed unneeded import

fickludd committed:
Improved thread-safety of background auth file read from disk

- now reads snapshots of the user and role filess until they are consistent, then atomically replaced user- and role repository contents in one go
- InternalFlatFileRealm now only starts read from disk on start, and stops reading on stop

davidegrohmann commented on community/security/src/main/java/org/neo4j/server/security/auth/FileRepositorySerializer.java:
```diff
@@ -34,34 +38,59 @@
+    public static byte[] readFromFile( FileSystemAbstraction fs, File file ) throws IOException
+    {
+        InputStream s = fs.openAsInputStream( file );
+        byte[] bytes = sun.misc.IOUtils.readFully( s, -1, false );
```
Did we test this on the IBM jdk?

davidegrohmann commented on community/security/src/main/java/org/neo4j/server/security/auth/FileRepositorySerializer.java:
```diff
@@ -34,34 +38,59 @@
-            Files.write( tempFile, serialize( records ) );
-            Files.move( tempFile, recordsFile, ATOMIC_MOVE, REPLACE_EXISTING );
+            writeToFile( fileSystem, tempFile, serialize( records ) );
+            fileSystem.move( tempFile, recordsFile, ATOMIC_MOVE, REPLACE_EXISTING );
```
What are we gonna do if `ATOMIC_MOVE` is unsupported?

fickludd replied:
We recover from transient errors during move, as tested in the FileUserRepository test. If neo4j is installed on a filesystem were an atomic move within the same directory is not possible then the flat-file based security is not recommended, and LDAP should be used instead.

fickludd committed:
Fixed private EphemeralFSA constructor and improved auth repo read and persist

fickludd committed:
Improved `Neo4jWithSocket` to be chainable with `EphemeralFileSystemRule`

- Fixed `BoltFullExchangesDocTest` to not mix ephemeral and real file systems

davidegrohmann commented on community/bolt/src/test/java/org/neo4j/bolt/v1/transport/integration/Neo4jWithSocket.java:
```diff
@@ -26,38 +26,48 @@
 
     public Neo4jWithSocket( TestGraphDatabaseFactory graphDatabaseFactory, Consumer<Map<Setting<?>, String>> configure )
     {
+        this( graphDatabaseFactory, () -> null, configure );
```
can we have a proper default here? What does `null` give you? "Ephemeral" or "real"?

fickludd replied:
Nice point, will fix as part of separate PR https://github.com/neo4j/neo4j/pull/7820, which needs to be rebased on this one.
